// @auto-it/core/dist/utils/logger.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.setLogLevel = exports.dummyLog = void 0;
const tslib_1 = require("tslib");
const signale_1 = tslib_1.__importDefault(require("signale"));
/** Create a dummy logger for testing. */
function dummyLog() {
    return {
        log: new signale_1.default.Signale({ disabled: true }),
        verbose: new signale_1.default.Signale({ disabled: true }),
        veryVerbose: new signale_1.default.Signale({ disabled: true }),
    };
}
exports.dummyLog = dummyLog;
const logger = {
    log: new signale_1.default.Signale(),
    verbose: new signale_1.default.Signale(),
    veryVerbose: new signale_1.default.Signale(),
};
/** Turn the logs on an off */
function toggleLogs(options) {
    Object.entries(options).forEach(([level, enabled]) => {
        if (enabled) {
            // @ts-ignore
            logger[level].enable();
        }
        else {
            // @ts-ignore
            logger[level].disable();
        }
    });
}
/** Set the log level */
function setLogLevel(newLogLevel) {
    logger.logLevel = newLogLevel;
    if (logger.logLevel === "verbose") {
        toggleLogs({ log: true, verbose: true, veryVerbose: false });
    }
    else if (logger.logLevel === "veryVerbose") {
        toggleLogs({ log: true, verbose: true, veryVerbose: true });
    }
    else if (logger.logLevel === "quiet") {
        toggleLogs({ log: false, verbose: false, veryVerbose: false });
    }
    else {
        toggleLogs({ log: true, verbose: false, veryVerbose: false });
    }
}
exports.setLogLevel = setLogLevel;
/** Create a logger the the given log level. */
function createLog() {
    return logger;
}
exports.default = createLog;
