// @auto-it/core/dist/utils/make-hooks.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.makeChangelogHooks = exports.makeInteractiveInitHooks = exports.makeLogParseHooks = exports.makeReleaseHooks = exports.makeHooks = void 0;
const tapable_1 = require("tapable");
/** Make the hooks for "auto" */
const makeHooks = () => ({
    beforeRun: new tapable_1.AsyncSeriesHook(["config"]),
    afterRun: new tapable_1.AsyncSeriesHook(["config"]),
    modifyConfig: new tapable_1.AsyncSeriesWaterfallHook(["config"]),
    validateConfig: new tapable_1.AsyncSeriesBailHook(["name", "options"]),
    beforeShipIt: new tapable_1.AsyncSeriesHook(["context"]),
    afterChangelog: new tapable_1.AsyncSeriesHook(["context"]),
    beforeCommitChangelog: new tapable_1.AsyncSeriesHook(["context"]),
    afterShipIt: new tapable_1.AsyncParallelHook(["context"]),
    makeRelease: new tapable_1.AsyncSeriesBailHook(["releaseInfo"]),
    afterRelease: new tapable_1.AsyncParallelHook(["releaseInfo"]),
    onCreateRelease: new tapable_1.SyncHook(["options"]),
    onCreateChangelog: new tapable_1.SyncHook(["changelog", "version"]),
    onCreateLogParse: new tapable_1.SyncHook(["logParse"]),
    getAuthor: new tapable_1.AsyncSeriesBailHook(),
    getPreviousVersion: new tapable_1.AsyncSeriesBailHook(),
    getRepository: new tapable_1.AsyncSeriesBailHook(),
    prCheck: new tapable_1.AsyncSeriesBailHook(["prInformation"]),
    beforeVersion: new tapable_1.AsyncSeriesHook(["context"]),
    version: new tapable_1.AsyncSeriesHook(["version"]),
    afterVersion: new tapable_1.AsyncSeriesHook(["context"]),
    publish: new tapable_1.AsyncSeriesHook(["version"]),
    afterPublish: new tapable_1.AsyncSeriesHook(),
    canary: new tapable_1.AsyncSeriesBailHook(["canaryContext"]),
    next: new tapable_1.AsyncSeriesWaterfallHook(["preReleaseVersions", "context"]),
});
exports.makeHooks = makeHooks;
/** Make the hooks for "Release" */
const makeReleaseHooks = () => ({
    onCreateChangelog: new tapable_1.SyncHook(["changelog", "version"]),
    createChangelogTitle: new tapable_1.AsyncSeriesBailHook([]),
    onCreateLogParse: new tapable_1.SyncHook(["logParse"]),
});
exports.makeReleaseHooks = makeReleaseHooks;
/** Make the hooks for "LogParse" */
const makeLogParseHooks = () => ({
    parseCommit: new tapable_1.AsyncSeriesWaterfallHook(["commit"]),
    omitCommit: new tapable_1.AsyncSeriesBailHook(["commit"]),
});
exports.makeLogParseHooks = makeLogParseHooks;
/** Make the hooks for "InteractiveInit" */
const makeInteractiveInitHooks = () => ({
    writeRcFile: new tapable_1.AsyncSeriesBailHook(["rcFile"]),
    getRepo: new tapable_1.AsyncSeriesBailHook([]),
    getAuthor: new tapable_1.AsyncSeriesBailHook([]),
    createEnv: new tapable_1.AsyncSeriesWaterfallHook(["vars"]),
    configurePlugin: new tapable_1.AsyncSeriesBailHook(["pluginName"]),
});
exports.makeInteractiveInitHooks = makeInteractiveInitHooks;
/** Make the hooks for "Changelog" */
const makeChangelogHooks = () => ({
    addToBody: new tapable_1.AsyncSeriesWaterfallHook(["notes", "commits"]),
    sortChangelogLines: new tapable_1.AsyncSeriesWaterfallHook(["lines"]),
    renderChangelogLine: new tapable_1.AsyncSeriesWaterfallHook(["line", "commit"]),
    renderChangelogTitle: new tapable_1.AsyncSeriesBailHook(["commits", "lineRender"]),
    renderChangelogAuthor: new tapable_1.AsyncSeriesBailHook([
        "author",
        "commit",
        "options",
    ]),
    renderChangelogAuthorLine: new tapable_1.AsyncSeriesBailHook(["author", "user"]),
    omitReleaseNotes: new tapable_1.AsyncSeriesBailHook(["commit"]),
});
exports.makeChangelogHooks = makeChangelogHooks;
